src/Control/Concurrent/CachedIO.hs

```diff
@@ -26,6 +27,11 @@ import Control.Monad.Catch (MonadCatch, onException)
 import Control.Monad.IO.Class (liftIO, MonadIO)
 import Data.Time.Clock (NominalDiffTime, addUTCTime, getCurrentTime, UTCTime)
 
+-- | A cached IO action in some monad @m@. Use 'runCached' to extract the action when you want to query it.
+--
+-- This newtype is intended to make it harder to accidentally 'Control.Monad.join' (and defeat all caching) when the fetch action and the outer monad are the same.
+newtype Cached t a = Cached {runCached :: t a}
```

Comment:
```suggestion
newtype Cached m a = Cached {runCached :: m a}
```
To match the use of `m` in the haddock; alternatively, call the type variable `m` in the haddock also.